heft-plugins/heft-sass-plugin/src/SassProcessor.ts

```diff
@@ -213,3 +229,69 @@ export class SassProcessor extends StringValuesTypingsGenerator {
     return url;
   }
 }
+
+interface IExtensionClassifierResult {
+  allFileExtensions: string[];
+  isFileModule: (relativePath: string) => boolean;
+}
+
+function buildExtensionClassifier(sassConfiguration: ISassConfiguration): IExtensionClassifierResult {
+  const {
+    fileExtensions: moduleFileExtensions = ['.sass', '.scss', '.css'],
+    nonModuleFileExtensions = ['.global.sass', '.global.scss', '.global.css']
+  } = sassConfiguration;
+
+  const hasModules: boolean = moduleFileExtensions.length > 0;
+  const hasNonModules: boolean = nonModuleFileExtensions.length > 0;
+
+  if (!hasModules) {
+    return {
+      allFileExtensions: nonModuleFileExtensions,
+      isFileModule: (relativePath: string) => false
+    };
+  }
+  if (!hasNonModules) {
+    return {
+      allFileExtensions: moduleFileExtensions,
+      isFileModule: (relativePath: string) => true
+    };
+  }
+
+  const extensionClassifier: Map<string, boolean> = new Map();
+  for (const extension of moduleFileExtensions) {
+    const normalizedExtension: string = extension.startsWith('.') ? extension : `.${extension}`;
+    extensionClassifier.set(normalizedExtension, true);
+  }
+
+  for (const extension of nonModuleFileExtensions) {
+    const normalizedExtension: string = extension.startsWith('.') ? extension : `.${extension}`;
+    const existingClassification: boolean | undefined = extensionClassifier.get(normalizedExtension);
+    if (existingClassification === true) {
+      throw new Error(
+        `File extension "${normalizedExtension}" is declared as both a SCSS module and not an SCSS module.`
+      );
+    }
+    extensionClassifier.set(normalizedExtension, false);
+  }
+
+  Sort.sortMapKeys(extensionClassifier, (key1, key2) => {
+    // Order by length, descending, so the longest gets tested first.
+    return key2.length - key1.length;
+  });
+
+  const isFileModule: (relativePath: string) => boolean = (relativePath: string) => {
+    // Naive comparison algorithm. O(E), where E is the number of extensions
+    // If performance becomes an issue, switch to using LookupByPath with a reverse iteration order using `.` as the delimiter
+    for (const [extension, isExtensionModule] of extensionClassifier) {
+      if (relativePath.endsWith(extension)) {
+        return isExtensionModule;
+      }
+    }
+    throw new Error(`Could not classify ${relativePath} as a SCSS module / not an SCSS module`);
```

Comment:
If no matches are found, this will throw. Is that expected? Will it only ever be provided SCSS files?

Reply:
Strictly speaking this error should be unreachable code, since the file extension lookup table here has the same set of extensions that are used to select the inputs to the typings generator. Yes, the function will only ever be provides SCSS (or CSS) files.